testnet/src/run_loop/neon.rs

```diff
@@ -41,64 +42,31 @@ impl RunLoop {
-                Some(mut tenure) => {
-                    self.callbacks.invoke_new_tenure(round_index, &burnchain_tip, &chain_tip, &mut tenure);
-                    tenure.run()
-                },
-                None => None
-            };
-
-            match artifacts_from_tenure {
-                Some(ref artifacts) => {
-                    node.commit_artifacts(
-                        &artifacts.anchored_block, 
-                        &artifacts.parent_block, 
-                        &mut burnchain, 
-                        artifacts.burn_fee);
-                },
-                None => {}
+            // (1) tell the relayer to check whether or not it won the sortition, and if so,
+            //     process and advertize the block
+            if !node.relayer_sortition_notify() {
+                // relayer hung up, exit.
+                error!("Block relayer and miner hung up, exiting.");
+                process::exit(1);
+            }
+
+            // (2) tell the relayer to run a new tenure
+            if !node.relayer_issue_tenure() {
+                // relayer hung up, exit.
+                error!("Block relayer and miner hung up, exiting.");
+                process::exit(1);
             }
 
             burnchain_tip = burnchain.sync();
-            self.callbacks.invoke_new_burn_chain_state(round_index, &burnchain_tip, &chain_tip);
-    
-            tenure = None;
-
-            // Have each node process the new block, that can include, or not, a sortition.
-            let (sortitioned_block, won_sortition) = node.process_burnchain_state(&burnchain_tip);
-
-            match (artifacts_from_tenure, sortitioned_block) {
-                // Pass if we're missing the artifacts from the current tenure.
-                (Some(ref artifacts), Some(ref last_sortitioned_block)) => {
-                    // Have each node process the previous tenure.
-                    // We should have some additional checks here, and ensure that the previous artifacts are legit.
-                    chain_tip = node.process_tenure(
-                        &artifacts.anchored_block, 
-                        &last_sortitioned_block.block_snapshot.burn_header_hash, 
-                        &last_sortitioned_block.block_snapshot.parent_burn_header_hash,             
-                        artifacts.microblocks.clone(),
-                        burnchain.burndb_mut());
-
-                        self.callbacks.invoke_new_stacks_chain_state(
-                            round_index, 
-                            &burnchain_tip, 
-                            &chain_tip, 
-                            &mut node.chain_state);
-                },
-                (_, _) => continue,
-            };
-            
-            // If the node we're looping on won the sortition, initialize and configure the next tenure
-            if won_sortition {
-                tenure = node.initiate_new_tenure();
-            } 
+
```

Comment:
Unrelated with the current line, but it looks like all the callbacks logic is gone.
We should not have contention on the callback struct so it can maybe be distributed across the threads behind a mutex, or even cloned.
I think that we can afford loosing the round_index (first argument shared by all the methods, since we have the block height) if it's an issue.

Reply:
Yep -- this disabled the callback logic for now, I can add back in, though.